File: src/muse/engine/providers/openai_provider.py
"""OpenAI GPT provider."""

import time

import openai

from muse.config import settings
from muse.engine.types import ModelResult


class OpenAIProvider:
    slug = "openai"
    name = "OpenAI GPT"

    async def generate(self, prompt: str, system: str, max_tokens: int = 2048) -> ModelResult:
        client = openai.AsyncOpenAI(api_key=settings.openai_api_key)
        start = time.perf_counter()
        response = await client.chat.completions.create(
            model=settings.openai_model,
            max_tokens=max_tokens,
            messages=[
                {"role": "system", "content": system},
                {"role": "user", "content": prompt},
            ],
        )
        latency = int((time.perf_counter() - start) * 1000)
        usage = response.usage
        return ModelResult(
            name=self.name,
            slug=self.slug,
            content=response.choices[0].message.content or "",
            input_tokens=usage.prompt_tokens if usage else 0,
            output_tokens=usage.completion_tokens if usage else 0,
            latency_ms=latency,
        )

    def is_available(self) -> bool:
        return bool(settings.openai_api_key)

File: MuseApp/MuseApp/Providers/OpenAIProvider.swift
// OpenAIProvider.swift — Direct OpenAI API calls (also works for OpenRouter)

import Foundation

class OpenAIProvider: ModelProvider {
    let slug: String
    let name: String
    private let model: String
    private let baseURL: String
    private let apiKeyName: String

    init(slug: String = "openai", name: String = "OpenAI GPT", model: String = "gpt-4o",
         baseURL: String = "https://api.openai.com/v1/chat/completions",
         apiKeyName: String = "openai_api_key") {
        self.slug = slug
        self.name = name
        self.model = model
        self.baseURL = baseURL
        self.apiKeyName = apiKeyName
    }

    func generate(prompt: String, system: String, maxTokens: Int) async throws -> ModelResult {
        let apiKey = KeychainManager.shared.get(key: apiKeyName)
        guard let apiKey, !apiKey.isEmpty else {
            return ModelResult(name: name, slug: slug, content: "", error: "API key not configured")
        }

        let messages = [
            ChatMessage(role: "system", content: system),
            ChatMessage(role: "user", content: prompt)
        ]

        let start = CFAbsoluteTimeGetCurrent()
        let result = try await APIClient.shared.chatCompletion(
            url: baseURL,
            apiKey: apiKey,
            model: model,
            messages: messages,
            maxTokens: maxTokens
        )
        let latency = Int((CFAbsoluteTimeGetCurrent() - start) * 1000)

        return ModelResult(
            name: name,
            slug: slug,
            content: result.content,
            inputTokens: result.inputTokens,
            outputTokens: result.outputTokens,
            latencyMs: latency
        )
    }

    func isAvailable() -> Bool {
        guard let key = KeychainManager.shared.get(key: apiKeyName) else { return false }
        return !key.isEmpty
    }
}
